pkg/providers/cloudstack/template.go

```diff
@@ -59,15 +60,19 @@ func getEtcdMachineSpec(clusterSpec v1alpha1.ClusterSpec, machineConfigs map[str
 func getWorkerNodeGroupMachineSpec(clusterSpec v1alpha1.ClusterSpec, machineConfigs map[string]*v1alpha1.CloudStackMachineConfig) (map[string]v1alpha1.CloudStackMachineConfigSpec, error) {
 	var workerNodeGroupMachineSpec *v1alpha1.CloudStackMachineConfigSpec
 	workerNodeGroupMachineSpecs := make(map[string]v1alpha1.CloudStackMachineConfigSpec, len(machineConfigs))
+	var allErrs []error
 	for _, wnConfig := range clusterSpec.WorkerNodeGroupConfigurations {
 		if wnConfig.MachineGroupRef != nil && machineConfigs[wnConfig.MachineGroupRef.Name] != nil {
 			workerNodeGroupMachineSpec = &machineConfigs[wnConfig.MachineGroupRef.Name].Spec
 			workerNodeGroupMachineSpecs[wnConfig.MachineGroupRef.Name] = *workerNodeGroupMachineSpec
+		} else {
+			allErrs = append(allErrs, errors.Errorf("machine group ref is not provided"))
 		}
 	}
 
-	if len(workerNodeGroupMachineSpecs) == 0 {
-		return workerNodeGroupMachineSpecs, errors.Errorf("getting worker node group configs")
+	if len(allErrs) > 0 {
+		aggregate := utilerrors.NewAggregate(allErrs)
+		return workerNodeGroupMachineSpecs, errors.Errorf("getting worker node group configs: %v", aggregate)
```

Comment:
Updating to adequately capture and error if any MachineGroupRef were missing. This would pass the case where there are multiple `WorkerNodeMachineGroupConfigurations` and any of them has a complete `MachineGroupRef` defined.